File: SimpleMoxieSwitcher-Windows/SimpleMoxieSwitcher/ViewModels/AllConversationsViewModel.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text.Json;
using System.Threading.Tasks;
using System.Windows.Input;
using System.Windows.Threading;
using SimpleMoxieSwitcher.Models;
using SimpleMoxieSwitcher.Services.Interfaces;

namespace SimpleMoxieSwitcher.ViewModels
{
    public class AllConversationsViewModel : INotifyPropertyChanged
    {
        private readonly IDockerService _dockerService;
        private DispatcherTimer _refreshTimer;

        private ObservableCollection<ConversationSession> _conversations = new ObservableCollection<ConversationSession>();
        private bool _isLoading = false;
        private string _errorMessage;

        public ObservableCollection<ConversationSession> Conversations
        {
            get => _conversations;
            set
            {
                _conversations = value;
                OnPropertyChanged();
            }
        }

        public bool IsLoading
        {
            get => _isLoading;
            set
            {
                _isLoading = value;
                OnPropertyChanged();
            }
        }

        public string ErrorMessage
        {
            get => _errorMessage;
            set
            {
                _errorMessage = value;
                OnPropertyChanged();
            }
        }

        // Commands
        public ICommand LoadConversationsCommand { get; }
        public ICommand StartAutoRefreshCommand { get; }
        public ICommand StopAutoRefreshCommand { get; }

        public AllConversationsViewModel(IDockerService dockerService)
        {
            _dockerService = dockerService;

            // Initialize commands
            LoadConversationsCommand = new RelayCommand(async () => await LoadConversations());
            StartAutoRefreshCommand = new RelayCommand(StartAutoRefresh);
            StopAutoRefreshCommand = new RelayCommand(StopAutoRefresh);
        }

        public async Task LoadConversations()
        {
            IsLoading = true;
            ErrorMessage = null;

            try
            {
                string pythonScript = @"
from hive.models import MoxieDevice, PersistentData
import json
from datetime import datetime

# Get the device
device = MoxieDevice.objects.filter(device_id='moxie_001').first()
if device:
    persist = PersistentData.objects.filter(device=device).first()
    if persist and persist.data:
        # Get real conversations
        real_convs = persist.data.get('real_conversations', [])

        # Group by date/session
        sessions = {}
        for conv in real_convs:
            # Extract date from timestamp
            timestamp = conv.get('timestamp', '')
            if timestamp:
                try:
                    dt = datetime.fromisoformat(timestamp.replace('Z', '+00:00'))
                    session_key = f""{dt.strftime('%Y-%m-%d_%H')}""

                    if session_key not in sessions:
                        sessions[session_key] = {
                            'id': session_key,
                            'personality': conv.get('personality', 'Default'),
                            'timestamp': timestamp,
                            'messages': []
                        }

                    # Add both user and moxie messages
                    if conv.get('user'):
                        sessions[session_key]['messages'].append({
                            'role': 'user',
                            'content': conv['user'],
                            'timestamp': timestamp
                        })
                    if conv.get('moxie'):
                        sessions[session_key]['messages'].append({
                            'role': 'moxie',
                            'content': conv['moxie'],
                            'timestamp': timestamp
                        })
                except:
                    pass

        # Convert to list and sort by timestamp
        session_list = list(sessions.values())
        session_list.sort(key=lambda x: x['timestamp'], reverse=True)

        # Output as JSON
        print(json.dumps(session_list))
    else:
        print(json.dumps([]))
else:
    print(json.dumps([]))
";

                var output = await _dockerService.ExecutePythonScript(pythonScript);

                // Parse the JSON output
                if (!string.IsNullOrEmpty(output))
                {
                    try
                    {
                        var sessionsArray = JsonSerializer.Deserialize<List<ConversationSessionData>>(output);

                        if (sessionsArray != null)
                        {
                            // Convert to our model
                            var conversations = sessionsArray.Select(sessionData =>
                            {
                                var messages = sessionData.Messages.Select(msg => new ChatMessage
                                {
                                    Role = msg.Role,
                                    Content = msg.Content,
                                    Timestamp = DateTime.TryParse(msg.Timestamp, out var dt) ? dt : DateTime.Now
                                }).ToList();

                                return new ConversationSession
                                {
                                    Id = sessionData.Id,
                                    Personality = sessionData.Personality,
                                    Timestamp = DateTime.TryParse(sessionData.Timestamp, out var dt) ? dt : DateTime.Now,
                                    Messages = messages,
                                    MessageCount = messages.Count
                                };
                            }).ToList();

                            // Update ObservableCollection
                            Conversations.Clear();
                            foreach (var conv in conversations)
                            {
                                Conversations.Add(conv);
                            }
                        }
                        else
                        {
                            Conversations.Clear();
                        }
                    }
                    catch (JsonException ex)
                    {
                        ErrorMessage = $"Failed to parse conversation data: {ex.Message}";
                        Conversations.Clear();
                    }
                }
                else
                {
                    // No real conversations yet
                    Conversations.Clear();
                }
            }
            catch (Exception ex)
            {
                ErrorMessage = $"Failed to load conversations: {ex.Message}";
                Conversations.Clear();
            }

            IsLoading = false;
        }

        public void StartAutoRefresh()
        {
            // Refresh every 5 seconds to show new conversations
            _refreshTimer = new DispatcherTimer
            {
                Interval = TimeSpan.FromSeconds(5)
            };
            _refreshTimer.Tick += async (sender, e) => await LoadConversations();
            _refreshTimer.Start();
        }

        public void StopAutoRefresh()
        {
            _refreshTimer?.Stop();
            _refreshTimer = null;
        }

        public event PropertyChangedEventHandler PropertyChanged;

        protected virtual void OnPropertyChanged([CallerMemberName] string propertyName = null)
        {
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
        }

        // Dispose pattern for timer cleanup
        public void Dispose()
        {
            _refreshTimer?.Stop();
            _refreshTimer = null;
        }
    }

    // Data structures for JSON parsing
    public class ConversationSessionData
    {
        public string Id { get; set; }
        public string Personality { get; set; }
        public string Timestamp { get; set; }
        public List<MessageData> Messages { get; set; }
    }

    public class MessageData
    {
        public string Role { get; set; }
        public string Content { get; set; }
        public string Timestamp { get; set; }
    }

    // Model for conversation sessions
    public class ConversationSession
    {
        public string Id { get; set; }
        public string Personality { get; set; }
        public DateTime Timestamp { get; set; }
        public List<ChatMessage> Messages { get; set; }
        public int MessageCount { get; set; }

        public string Preview
        {
            get
            {
                var firstUserMessage = Messages?.FirstOrDefault(m => m.Role == "user");
                if (firstUserMessage != null)
                {
                    var preview = firstUserMessage.Content;
                    return preview.Length > 100 ? preview.Substring(0, 100) + "..." : preview;
                }
                return "No messages";
            }
        }

        public string FormattedDate
        {
            get
            {
                return Timestamp.ToString("MMM dd, yyyy h:mm tt");
            }
        }
    }

    public class ChatMessage
    {
        public string Role { get; set; }
        public string Content { get; set; }
        public DateTime Timestamp { get; set; }
    }
}

File: SimpleMoxieSwitcher/Sources/SimpleMoxieSwitcher/ViewModels/AllConversationsViewModel.swift
import Foundation
import SwiftUI

@MainActor
class AllConversationsViewModel: ObservableObject {
    @Published var conversations: [ConversationSession] = []
    @Published var isLoading = false
    @Published var errorMessage: String?

    private let dockerService: DockerServiceProtocol
    private var refreshTimer: Timer?

    init(dockerService: DockerServiceProtocol) {
        self.dockerService = dockerService
    }

    func loadConversations() async {
        isLoading = true
        errorMessage = nil

        do {
            let pythonScript = """
            from hive.models import MoxieDevice, PersistentData
            import json
            from datetime import datetime

            # Get the device
            device = MoxieDevice.objects.filter(device_id='moxie_001').first()
            if device:
                persist = PersistentData.objects.filter(device=device).first()
                if persist and persist.data:
                    # Get real conversations
                    real_convs = persist.data.get('real_conversations', [])

                    # Group by date/session
                    sessions = {}
                    for conv in real_convs:
                        # Extract date from timestamp
                        timestamp = conv.get('timestamp', '')
                        if timestamp:
                            try:
                                dt = datetime.fromisoformat(timestamp.replace('Z', '+00:00'))
                                session_key = f"{dt.strftime('%Y-%m-%d_%H')}"

                                if session_key not in sessions:
                                    sessions[session_key] = {
                                        'id': session_key,
                                        'personality': conv.get('personality', 'Default'),
                                        'timestamp': timestamp,
                                        'messages': []
                                    }

                                # Add both user and moxie messages
                                if conv.get('user'):
                                    sessions[session_key]['messages'].append({
                                        'role': 'user',
                                        'content': conv['user'],
                                        'timestamp': timestamp
                                    })
                                if conv.get('moxie'):
                                    sessions[session_key]['messages'].append({
                                        'role': 'moxie',
                                        'content': conv['moxie'],
                                        'timestamp': timestamp
                                    })
                            except:
                                pass

                    # Convert to list and sort by timestamp
                    session_list = list(sessions.values())
                    session_list.sort(key=lambda x: x['timestamp'], reverse=True)

                    # Output as JSON
                    print(json.dumps(session_list))
                else:
                    print(json.dumps([]))
            else:
                print(json.dumps([]))
            """

            let output = try await dockerService.executePythonScript(pythonScript)

            // Parse the JSON output
            if let jsonData = output.data(using: .utf8),
               let sessionsArray = try? JSONDecoder().decode([ConversationSessionData].self, from: jsonData) {

                // Convert to our model
                conversations = sessionsArray.map { sessionData in
                    let messages = sessionData.messages.map { msg in
                        ChatMessage(
                            role: msg.role,
                            content: msg.content,
                            timestamp: ISO8601DateFormatter().date(from: msg.timestamp) ?? Date()
                        )
                    }

                    return ConversationSession(
                        id: sessionData.id,
                        personality: sessionData.personality,
                        timestamp: ISO8601DateFormatter().date(from: sessionData.timestamp) ?? Date(),
                        messages: messages,
                        messageCount: messages.count
                    )
                }
            } else {
                // No real conversations yet - clear fake data
                conversations = []
            }

        } catch {
            errorMessage = "Failed to load conversations: \(error.localizedDescription)"
            conversations = []
        }

        isLoading = false
    }

    func startAutoRefresh() {
        // Refresh every 5 seconds to show new conversations
        refreshTimer = Timer.scheduledTimer(withTimeInterval: 5.0, repeats: true) { [weak self] _ in
            Task { @MainActor [weak self] in
                await self?.loadConversations()
            }
        }
    }

    func stopAutoRefresh() {
        refreshTimer?.invalidate()
        refreshTimer = nil
    }

    deinit {
        refreshTimer?.invalidate()
    }
}

// Data structures for JSON parsing
struct ConversationSessionData: Codable {
    let id: String
    let personality: String
    let timestamp: String
    let messages: [MessageData]
}

struct MessageData: Codable {
    let role: String
    let content: String
    let timestamp: String
}

// Model for conversation sessions
struct ConversationSession: Identifiable {
    let id: String
    let personality: String
    let timestamp: Date
    let messages: [ChatMessage]
    let messageCount: Int

    var preview: String {
        if let firstUserMessage = messages.first(where: { $0.role == "user" }) {
            let preview = firstUserMessage.content
            return String(preview.prefix(100)) + (preview.count > 100 ? "..." : "")
        }
        return "No messages"
    }

    var formattedDate: String {
        let formatter = DateFormatter()
        formatter.dateStyle = .medium
        formatter.timeStyle = .short
        return formatter.string(from: timestamp)
    }
}